src/util_i_variables.nss

```diff
@@ -0,0 +1,2051 @@
+    if (nTime != 0)     SqlBindInt   (q, "@time", abs(nTime));
+    return q;
+}
+
+/// @brief Prepares a complex select query to retrieve multiple variables by criteria.
+/// @param nType Bitwise VARIABLE_TYPE_*
+/// @param sVarName Variable name pattern, accept glob patterns, sets and wildcards
+/// @param sTag Tag pattern, accepts glob patterns, sets and wildcards
+/// @param nTime (+) is after, (-) is before
+/// @warning If no parameters are passed, this query will result in no variables being
+///     retrieved.
+sqlquery _PrepareComplexVariableSelect(object oObject, int nType = VARIABLE_TYPE_NONE, string sVarName = "", 
+                                       string sTag = "", int nTime = 0, int bCampaign = FALSE)
+{
+    int n, bPC = GetIsPC(oObject);
+    string sTable = bPC ? VARIABLE_TABLE_PC : VARIABLE_TABLE_MODULE;
+    string sWhere =  (sVarName == "" ? "" : " $" + IntToString(++n) + " varname GLOB @varname");
+           sWhere += (sTag == ""     ? "" : " $" + IntToString(++n) + " tag GLOB @tag");
+           sWhere += (nType < 0      ? "" : " $" + IntToString(++n) + " type & @type > 0");
+           sWhere += (nTime == 0     ? "" : " $" + IntToString(++n) + " timestamp " + (nTime > 0 ? ">" : "<") + " @time");
+    
+    json jKeyWords = ListToJson("WHERE,AND,AND,AND");
+    string s = "SELECT type, varname, value, tag, timestamp FROM " + sTable + sWhere + ";";
+           s = SubstituteString(s, jKeyWords);
+
+    sqlquery q = bPC || !bCampaign ? _PrepareQueryObject(oObject, s) : _PrepareQueryCampaign(s);
+    if (sVarName != "") SqlBindString(q, "@varname", sVarName);
+    if (sTag != "")     SqlBindString(q, "@tag", sTag);
+    if (nType > 0)      SqlBindInt   (q, "@type", nType);
+    if (nTime != 0)     SqlBindInt   (q, "@time", abs(nTime));
+    return q;
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteAll(object oObject, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, 0, "", sTag, 0, bCampaign);
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteByName(object oObject, string sVarName, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, 0, sVarName, sTag, 0, bCampaign);
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteByType(object oObject, int nType, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, nType, "", sTag, 0, bCampaign);
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteByTime(object oObject, int nTime, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, 0, "", sTag, nTime, bCampaign);
+}
+
+/// @brief Increments/Decremenst an existing variable (int/float).  If the variable
+///     does not exist, creates variables, then increments/decrements.
+sqlquery _PrepareVariableIncrement(object oObject, string sVarName, string sTag, int bCampaign)
```

Comment:
Why is setting `@type` handled in the calling function and not in here?

Reply:
Modified via commit.